ethereum/eip712/encoding.go

```diff
@@ -116,11 +117,31 @@ func decodeAminoSignDoc(signDocBytes []byte) (apitypes.TypedData, error) {
 		return apitypes.TypedData{}, err
 	}
 
+	// Cross chain signing portion to convert aminoDoc.ChainId to the chainID of the signed chain
+	memo := aminoDoc.Memo
+	if strings.Contains(memo, "|CROSSCHAIN-SIGNING|") {
+		splitMemoCrossChain := strings.Split(memo, "|CROSSCHAIN-SIGNING|")
+		if len(splitMemoCrossChain) != 2 {
+			return apitypes.TypedData{}, errors.New("invalid memo")
+		}
+		memoSuffix := splitMemoCrossChain[1]
+		memoChainIDs := strings.Split(memoSuffix, ";")
+
+		if len(memoChainIDs) != 2 {
+			return apitypes.TypedData{}, errors.New("invalid memo")
+		}
+		signedChainSplit := strings.Split(memoChainIDs[0], ":")
+		carbonChainSplit := strings.Split(memoChainIDs[1], ":")
+		if len(signedChainSplit) != 2 || len(carbonChainSplit) != 2 {
+			return apitypes.TypedData{}, errors.New("invalid memo")
+		}
+		aminoDoc.ChainID = signedChainSplit[1]
```

Comment:
i think we can shift the replay attack guard here, since the chainId is just a static variable initialised in Makefile.

import the evm keeper:

```
evmkeeper "github.com/evmos/ethermint/x/evm/keeper"
```

// Check the carbon chain ID to prevent replay attack
```
				if carbonChainSplit[1] != evmkeeper.EvmChainId {
					return apitypes.TypedData{}, fmt.Errorf("invalid chainId, expected %v, got %v", evmkeeper.EvmChainId, carbonChainSplit[1])
				}
```

Reply:
Done in 113d5d63f87b767bf5b8a4ffbe7153ab4dec46e3